arealite/impl/trainer/grpo.py

```diff
@@ -0,0 +1,530 @@
+        else:
+            world_size = 1
+        global_step = 0
+        warmup_steps = self.max_head_offpolicyness + 1
+        assert steps_per_epoch >= warmup_steps
+        start_time = time.monotonic()
+        for epoch in range(total_epochs):
+            for step, data in enumerate(self.train_dataloader):
+                timing_stats = {}
+                with record_timing("timeperf/rollout", timing_stats):
+                    if self.config.async_training:
+                        self.rollout_controller.submit(data)
+                        # Submitted data will not actually be sent for rollout.
+                        # The rollout controller over-subscribe the data to
+                        # ensure that there are enough data being generated.
+                        if epoch == 0 and step < warmup_steps:
+                            continue
+                        # Wait until enough trajectories has been collected.
+                        trajs = self.rollout_controller.prepare_batch(
+                            batch_size=self.args.train_dataset.batch_size // world_size
+                        )
+                    else:
+                        # Run batched rollout by submitting requests to LLM servers
+                        trajs = self.rollout_controller.generate_batch(
+                            batch_size=len(data),
+                            env_options=data,
+                        )
+
+                with record_timing("timeperf/train_step", timing_stats):
+                    # Run RL training and update weights.
+                    mb_stats = self._train_step(trajs)
+                    self.actor.step_lr_scheduler()
+
+                with record_timing("timeperf/sync_weights", timing_stats):
+                    # Synchronize weights to the client.
+                    self.actor.update_weights_to(self.llm_client)
+                    # Update model version
+                    name = names.model_version(
+                        self.args.experiment_name, self.args.trial_name, "actor"
+                    )
+                    name_resolve.add(name, str(global_step + 1), replace=True)
+
+                if self.save_ctl.check(
+                    epochs=int(step == steps_per_epoch - 1), steps=1
+                ):
+                    if dist.get_rank() == 0:
+                        logger.info("Saving model ...")
+                    with record_timing("timeperf/save", timing_stats):
+                        save_path = os.path.join(
+                            constants.get_save_path(self.args), "actor"
+                        )
+                        self.actor.save_model_to_hf(
+                            save_path,
+                            tokenizer=self.actor_tokenizer,
+                            base_model_path=self.config.actor.path,
+                        )
+
+                assert len(mb_stats) == self.config.ppo_n_minibatches
```

Comment:
Throw some hints here such as f"Micro batch mismatch, current {mb_stats} and config {self.config.ppo_n_minibatches}. Check your configuration. "